frb_codegen/src/commands.rs

```diff
@@ -18,31 +21,32 @@ fn call_shell(cmd: &str) -> Output {
 }
 
 pub fn ensure_tools_available(dart_root: &str) -> Result {
-    let toolchain = guess_toolchain(dart_root).unwrap();
-    if toolchain == DartToolchain::Dart && !call_shell("dart --version").status.success() {
-        return Err(Error::MissingExe("dart".to_string()));
-    } else if toolchain == DartToolchain::Flutter
-        && !call_shell("flutter --version").status.success()
-    {
-        return Err(Error::MissingExe("flutter".to_string()));
+    let repo =
+        DartRepository::from_str(dart_root).map_err(|e| Error::StringError(e.to_string()))?;
+    if !repo.toolchain_available() {
+        return Err(Error::MissingExe(repo.toolchain.to_string()));
     }
 
-    let ffi = ensure_dependencies(dart_root, "ffi", PackageManager::Dependencies).unwrap();
-    if ffi.is_none() {
-        return Err(Error::MissingDep {
-            name: "ffi".into(),
-            context: PackageManager::Dependencies,
-            version: "^2.0.1".into(),
-        });
+    if let Err(e) = repo.has_specified("ffi", PackageManager::Dependencies, &FFI_REQUIREMENT) {
+        return Err(e.into());
+    }
```

Comment:
Seems Rust has an operator `?` for this usage: `some_function_that_return_result()?` means `if error return error`

p.s. the same applies for 4 places around this code

Reply:
done in d7f913f9665145617c107e8a4d529226be3b665e